tests/slate/test_assemble_tensors.py

```diff
@@ -170,6 +170,76 @@ def test_mixed_argument_tensor(mesh):
     assemble(T)
 
 
+def test_vector_subblocks(mesh):
+    V = VectorFunctionSpace(mesh, "DG", 3)
+    U = FunctionSpace(mesh, "DG", 2)
+    T = FunctionSpace(mesh, "DG", 1)
+    W = V * U * T
+    x = SpatialCoordinate(mesh)
+    q = Function(V).project(grad(sin(pi*x[0])*cos(pi*x[1])))
+    p = Function(U).interpolate(-x[0]*exp(-x[1]**2))
+    r = Function(T).assign(42.0)
+    u, phi, eta = TrialFunctions(W)
+    v, psi, nu = TestFunctions(W)
+
+    K = Tensor(inner(u, v)*dx + inner(phi, psi)*dx + inner(eta, nu)*dx)
+    F = Tensor(inner(q, v)*dx + inner(p, psi)*dx + inner(r, nu)*dx)
+    E = K.inv * F
+    Eq = assemble(E.block(0))
+    Ep = assemble(E.block(1))
+    Er = assemble(E.block(2))
+
+    assert np.allclose(Eq.dat.data, q.dat.data, rtol=1e-14)
+    assert np.allclose(Ep.dat.data, p.dat.data, rtol=1e-14)
+    assert np.allclose(Er.dat.data, r.dat.data, rtol=1e-14)
+
+
+def test_matrix_subblocks(mesh):
+    if mesh.ufl_cell() == quadrilateral:
+        U = FunctionSpace(mesh, "RTCF", 2)
+    else:
+        U = FunctionSpace(mesh, "RT", 2)
+    V = FunctionSpace(mesh, "DG", 1)
+    T = FunctionSpace(mesh, "HDiv Trace", 1)
+    n = FacetNormal(mesh)
+    W = U * V * T
+    u, p, lambdar = TrialFunctions(W)
+    w, q, gammar = TestFunctions(W)
+
+    A = Tensor(inner(u, w)*dx + p*q*dx - div(w)*p*dx + q*div(u)*dx +
+               lambdar('+')*jump(w, n=n)*dS + gammar('+')*jump(u, n=n)*dS +
+               lambdar*gammar*ds)
+    A00 = assemble(A.block(0, 0))
+    A01 = assemble(A.block(0, 1))
+    A02 = assemble(A.block(0, 2))
+    A10 = assemble(A.block(1, 0))
+    A11 = assemble(A.block(1, 1))
+    A20 = assemble(A.block(2, 0))
+    A22 = assemble(A.block(2, 2))
+
+    u = TrialFunction(U)
+    w = TestFunction(U)
+    p = TrialFunction(V)
+    q = TestFunction(V)
+    lambdar = TrialFunction(T)
+    gammar = TestFunction(T)
+    A00ref = assemble(inner(u, w)*dx)
+    A01ref = assemble(-div(w)*p*dx)
+    A02ref = assemble(lambdar('+')*jump(w, n=n)*dS)
+    A10ref = assemble(q*div(u)*dx)
+    A11ref = assemble(p*q*dx)
+    A20ref = assemble(gammar('+')*jump(u, n=n)*dS)
+    A22ref = assemble(lambdar*gammar*ds)
+
+    assert np.allclose(A00.M.values, A00ref.M.values, rtol=1e-14)
+    assert np.allclose(A01.M.values, A01ref.M.values, rtol=1e-14)
+    assert np.allclose(A02.M.values, A02ref.M.values, rtol=1e-14)
+    assert np.allclose(A10.M.values, A10ref.M.values, rtol=1e-14)
+    assert np.allclose(A11.M.values, A11ref.M.values, rtol=1e-14)
+    assert np.allclose(A20.M.values, A20ref.M.values, rtol=1e-14)
+    assert np.allclose(A22.M.values, A22ref.M.values, rtol=1e-14)
```

Comment:
Can you test that taking a `block` resulting in a mixed matrix, and then taking a block of that (into a non-mixed matrix) and then assembling that works too?

Reply:
I note in passing that if you made the API consistent between form-splitting and tensor blocking, this test could be much reduced in length.